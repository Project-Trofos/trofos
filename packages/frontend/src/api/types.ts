export type Project = {
  id: number;
  pname: string;
  pkey: string | null;
  description: string | null;
  course_id: number | null;
  public: boolean;
  created_at: string;
  course?: Course;
  is_archive: boolean | null;
};

export type Course = {
  id: number;
  code: string;
  startYear: number;
  startSem: number;
  endYear: number;
  endSem: number;
  cname: string;
  description: string | null;
  public: boolean;
  created_at: string;
  shadow_course: boolean;
  is_archive: boolean | null;
};

export type User = {
  user_email: string;
  user_display_name: string;
  user_id: number;
  projects: Project[];
  courses: CourseRoles[];
  basicRoles: BasicRoles[];
};

export type BasicRoles = {
  user_email: string;
  role_id: number;
};

export type CourseRoles = {
  id: number;
  user_email: string;
  role_id: number;
  course_id: number;
};

export type CreateUserRequest = {
  userEmail: string;
  newPassword: string;
};

export type RegisterUser = {
  userEmail: string;
  newPassword: string;
  userDisplayName: string;
};

export type UserCourseRoleRequest = {
  id: number;
  userRole: number;
  userId: number;
};

export type Role = {
  role_name: string;
  id: number;
};

export type Action = {
  action: string;
};

export type ActionOnRole = {
  id: number;
  action: string;
};

export type ActionsOnRoles = {
  id: number;
  role_name: string;
  actions: Action[];
};

export type ProjectData = Project & {
  course: Course | null;
  users: UserData[];
  sprints: {
    id: number;
    name: string;
  }[];
  backlogStatuses: Omit<BacklogStatusData, 'project_id'>[];
  telegramChannelLink: string;
  owner_id: number | null;
};

export type UserData = {
  user: Pick<User, 'user_id' | 'user_email' | 'user_display_name' | 'courses'>;
};

export type ScrumBoardUserData = {
  user: {
    user_id: number | null;
    user_display_name: string;
    user_email: string;
  };
};

export type Milestone = {
  id: number;
  name: string;
  start_date: string;
  deadline: string;
  course_id: number;
  created_at: string;
};

export type Announcement = {
  id: number;
  title: string;
  content: string;
  user_id: number;
  created_at: string;
  updated_at?: string;
};

// CourseData type returned by BE api. We perform transformResponse to convert the data into type CourseData
export type CourseDataResponse = Omit<CourseData, 'users'> & {
  courseRoles: UserData[];
};

export type CourseData = Course & {
  milestones: Milestone[];
  announcements: Announcement[];
  users: UserData[];
};

export type UserOnRolesOnCourse = {
  id: number;
  user_id: number;
  role_id: number;
  course_id: number;
  role: {
    id: number;
    role_name: string;
  };
};

export type BacklogStatusData = {
  name: string;
  type: 'todo' | 'in_progress' | 'done';
  project_id: number;
  order: number;
};

export type Comment = {
  comment_id: number;
  backlog_id: number;
  project_id: number;
  commenter_id: number;
  content: string;
  created_at: string;
  updated_at: string | null;
  commenter: {
    created_at: string;
    project_id: string;
    user: {
      user_id: number;
      user_email: string;
      user_display_name: string;
    };
  };
};

export type CommentFieldsType = {
  projectId: number;
  backlogId: number;
  commenterId: number;
  content: string;
};

export enum BacklogStatus {
  TODO = 'To do',
  DONE = 'Done',
}

export type BacklogPriority = 'very_high' | 'high' | 'medium' | 'low' | 'very_low' | null;

export type Backlog = {
  backlog_id: number;
  summary: string;
  type: 'story' | 'task' | 'bug';
  priority: BacklogPriority;
  reporter_id: number;
  assignee_id: number | null;
  sprint_id: number | null;
  points: number | null;
  description: string | null;
  project_id: number;
  status: BacklogStatus | string;
  assignee: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
  epic?: Epic | null;
};

export type Epic = {
  epic_id: number;
  project_id: number;
  name: string;
  description: string | null;
};

export type DefaultBacklog = {
  backlog_id?: number;
  summary?: string;
  type?: 'story' | 'task' | 'bug';
  priority?: BacklogPriority;
  reporter_id?: number;
  assignee_id?: number | null;
  sprint_id?: number | null;
  points?: number | null;
  description?: string | null;
  project_id?: number;
  status?: BacklogStatus | string;
  assignee?: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
};

export type BacklogUpdatePayload = {
  projectId: number;
  backlogId: number;
  srcSprintId?: number | null;
  fieldToUpdate: Partial<Backlog>;
};

export enum BacklogHistoryType {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
}

export type BacklogHistory = {
  history_type: BacklogHistoryType;
  date: string;
} & Omit<Backlog, 'assignee' | 'summary' | 'description'>;

export type UpdateUserRolePayload = {
  userId: number;
  newRoleId: number;
};

export type ProjectGitLink = {
  project_id: number;
  repo: string;
};

export type ProjectGitLinkData = {
  projectId: number;
  repoLink: string;
};

export type ProjectUserSettings = {
  project_id: number;
  user_id: number;
  email_notification: boolean;
};

export type Settings = {
  current_year: number;
  current_sem: number;
};

export type RetrospectiveVote = {
  id: number;
  retro_id: number;
  user_id: number;
  type: RetrospectiveVoteType;
};

export type Retrospective = {
  id: number;
  sprint_id: number;
  content: string;
  type: RetrospectiveType;
  score: number;
  votes: RetrospectiveVote[];
  is_action_taken: boolean;
};

export enum RetrospectiveType {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  ACTION = 'action',
}

export enum RetrospectiveVoteType {
  UP = 'up',
  DOWN = 'down',
}

export type CourseImportCsvPayload = {
  courseId: number;
  payload: FormData;
};

export type Feedback = {
  id: number;
  user_id?: number;
  user: {
    user_email: string;
    user_display_name: string;
  };
  sprint_id: number;
  content: string;
  created_at: string;
  updated_at?: string;
};

export type OAuth2Payload = {
  code: string;
  state: string;
  callbackUrl: string;
};

export type Invite = {
  project_id: number;
  email: string;
  expiry_date: Date;
};

export type UserApiKey = {
  id: number;
  user_id: number;
  api_key: string;
  created_at: string;
  last_used: string | null;
  active: boolean;
};

export type UserGuideQueryResponse = {
  answer: string;
  links: Array<string>;
};

export type FeatureFlag = {
  feature_name: string;
  active: boolean;
};

export type ProjectAssignment = {
  id: number;
  targetProjectId: number;
  sourceProjectId: number;
};
