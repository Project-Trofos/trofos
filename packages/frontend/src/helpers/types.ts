export type PickRename<T, K extends keyof T, R extends PropertyKey> = Pick<Omit<T, K> & { [P in R]: T[K] }, R>;
