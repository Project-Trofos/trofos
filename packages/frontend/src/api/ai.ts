import trofosApiSlice from '.';
import { UserGuideQueryResponse, UserGuideRecommendation } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    answerUserGuideQuery: builder.mutation<UserGuideQueryResponse, { query: string; isEnableMemory: boolean }>({
      query: ({ query, isEnableMemory }) => ({
        url: `ai/userGuideQuery`,
        credentials: 'include',
        method: 'POST',
        body: { query, isEnableMemory },
      }),
    }),
    recommendUserGuideSections: builder.mutation<UserGuideRecommendation[], void>({
      query: () => ({
        url: `ai/recommendUserGuide`,
        method: 'POST',
        credentials: 'include',
      }),
    }),
  }),
});

export const { useAnswerUserGuideQueryMutation, useRecommendUserGuideSectionsMutation } = extendedApi;
