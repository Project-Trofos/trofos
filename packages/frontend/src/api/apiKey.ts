import trofosApiSlice from '.';
import { UserApiKey } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getUserApiKey: builder.query<UserApiKey, void>({
      query: () => ({
        url: 'api-key/me',
        credentials: 'include',
      }),
      providesTags: ['UserApiKey'],
    }),
    generateUserApiKey: builder.mutation<UserApiKey, void>({
      query: () => ({
        url: 'api-key/generate',
        method: 'POST',
        credentials: 'include',
      }),
      invalidatesTags: ['UserApiKey'],
    }),
  }),
});

export const { useGetUserApiKeyQuery, useGenerateUserApiKeyMutation } = extendedApi;
