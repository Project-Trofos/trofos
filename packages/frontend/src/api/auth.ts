import trofosApiSlice from '.';

export type UserLoginInfo = {
  userEmail: string,
  userPassword: string
};

export type UserInfo = {
  userEmail: string,
  userRole: number,
  userId: number
};

export type ChangePassword = {
  userId : number | undefined,
  oldUserPassword : string,
  newUserPassword: string,
}

// Auth APIs
const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    loginUser: builder.mutation<void, UserLoginInfo>({
      query: (userLoginInfo) => ({
        url: '/account/login',
        method: 'POST',
        body : {
          userEmail: userLoginInfo.userEmail,
          userPassword: userLoginInfo.userPassword,
        },
        credentials: 'include',
      }),
    }),
    logoutUser: builder.mutation<void, void>({
      query: () => ({
        url: '/account/logout',
        method: 'POST',
        credentials: 'include',
      }),
      invalidatesTags : ['UserInfo'],
    }),
    getUserInfo: builder.query<UserInfo, void>({
      query: () => ({
        url: '/account/userInfo',
        credentials: 'include',
      }),
      providesTags: ['UserInfo'],
    }),
    changePassword : builder.mutation<void, ChangePassword>({
      query: (changePassword) => ({
        url: '/account/changePassword',
        method: 'POST',
        body : {
          userId : changePassword.userId,
          oldUserPassword : changePassword.oldUserPassword,
          newUserPassword : changePassword.newUserPassword
        },
        credentials: 'include'
      })
    })
  }),
});

export const { 
  useLoginUserMutation, 
  useLogoutUserMutation,
  useGetUserInfoQuery,
  useChangePasswordMutation
} = extendedApi;