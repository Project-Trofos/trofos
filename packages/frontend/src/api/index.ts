import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Define a service using a base URL and expected endpoints
const trofosApiSlice = createApi({
  reducerPath: 'trofosApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NODE_ENV === 'test' ? 'http://localhost:3001/api' : '/api',
    prepareHeaders: (headers) => {
      headers.set('Cache-control', 'no-cache');
      return headers;
    },
  }),
  tagTypes: [
    'Project',
    'UserInfo',
    'Course',
    'Backlog',
    'BacklogHistory',
    'User',
    'Sprint',
    'ActionsOnRoles',
    'CourseRoles',
    'ProjectRoles',
    'Settings',
    'StandUpHeader',
    'StandUp',
    'Retrospective',
    'ProjectUserSettings',
    'Feedback',
  ],
  endpoints: () => ({}),
});

// Define a service using a base URL and expected endpoints
export const nusmodsApiSlice = createApi({
  reducerPath: 'nusmodsApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.nusmods.com/v2/2022-2023' }),
  endpoints: () => ({}),
});

export default trofosApiSlice;
