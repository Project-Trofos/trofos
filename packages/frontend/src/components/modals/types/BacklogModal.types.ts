export interface BacklogFormFields extends FormData {
  projectId: number;
}

export type BacklogSelect = {
  id: string;
  name: string;
};
