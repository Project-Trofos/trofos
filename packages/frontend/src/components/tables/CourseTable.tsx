import React from 'react';
import { Card, Space, Table } from 'antd';
import { Link } from 'react-router-dom';
import { Course } from '../../api/types';
import { Subheading } from '../typography';

import './ProjectTable.css';
import { filterDropdown } from './helper';

type CourseTableProps = {
  courses: Course[] | undefined;
  isLoading: boolean;
  heading?: string;
  control?: React.ReactNode;
};

/**
 * Table for listing projects
 */
export default function CourseTable({ courses: projects, isLoading, heading, control }: CourseTableProps) {
  return (
    <Card className="table-card">
      <Space direction="vertical" size="middle" style={{ display: 'flex' }}>
        <Space style={{ display: 'flex', justifyContent: 'space-between' }}>
          <Subheading>{heading ?? 'Courses'}</Subheading>
          {control}
        </Space>
        <Table
          dataSource={projects}
          rowKey={(project) => project.id}
          loading={isLoading}
          bordered
          size="small"
          pagination={{ pageSize: 5 }}
        >
          <Table.Column
            width={150}
            title="ID"
            dataIndex="id"
            sorter={(a: Course, b: Course) => a.id.localeCompare(b.id)}
          />
          <Table.Column width={150} title="Year" dataIndex="year" sorter={(a: Course, b: Course) => a.year - b.year} />
          <Table.Column width={150} title="Semester" dataIndex="sem" sorter={(a: Course, b: Course) => a.sem - b.sem} />
          <Table.Column
            title="Name"
            dataIndex="cname"
            filterDropdown={filterDropdown}
            sorter={(a: Course, b: Course) => a.cname.localeCompare(b.cname)}
            onFilter={(value, record: Course) => record.cname.toLowerCase().includes(value.toString().toLowerCase())}
          />
          <Table.Column
            width={150}
            title="Action"
            dataIndex="action"
            render={(_, record: Course) => (
              <Space size="middle">
                <Link to={`/course/${record.id}/overview`}>Go to</Link>
              </Space>
            )}
          />
        </Table>
      </Space>
    </Card>
  );
}
