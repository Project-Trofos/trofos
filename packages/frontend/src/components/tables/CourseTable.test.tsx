import React from 'react';
import { render, screen } from '@testing-library/react';

import { BrowserRouter } from 'react-router-dom';
import CourseTable from './CourseTable';
import { Course } from '../../api/types';

describe('test CourseTable', () => {
  const mockCourses: Course[] = [
    {
      id: 'course_id',
      year: 2022,
      sem: 10,
      cname: 'course1',
      description: 'project1_description',
      public: false,
      created_at: '2022-09-15T01:58:01.735Z',
    },
  ];

  const setup = () => {
    const { baseElement, debug } = render(
      <BrowserRouter>
        <CourseTable courses={mockCourses} isLoading={false} />
      </BrowserRouter>,
    );
    return { baseElement, debug };
  };

  it('should render table with correct fields', () => {
    const { baseElement } = setup();

    expect(screen.getByText('Courses')).toBeInTheDocument();

    // Ensure columns are present
    expect(screen.getByText('ID')).toBeInTheDocument();
    expect(screen.getByText('Name')).toBeInTheDocument();
    expect(screen.getByText('Action')).toBeInTheDocument();

    // Ensure row is present
    expect(screen.getByText(mockCourses[0].id)).toBeInTheDocument();
    expect(screen.getByText(mockCourses[0].year)).toBeInTheDocument();
    expect(screen.getByText(mockCourses[0].sem)).toBeInTheDocument();
    expect(screen.getByText(mockCourses[0].cname)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });
});
