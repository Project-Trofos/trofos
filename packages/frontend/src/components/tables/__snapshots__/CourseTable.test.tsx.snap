// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test CourseTable should render table with correct fields 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered table-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="ant-space ant-space-vertical"
          style="display: flex;"
        >
          <div
            class="ant-space-item"
            style="margin-bottom: 16px;"
          >
            <div
              class="ant-space ant-space-horizontal ant-space-align-center"
              style="display: flex; justify-content: space-between;"
            >
              <div
                class="ant-space-item"
              >
                <h4
                  class="ant-typography"
                >
                  Courses
                </h4>
              </div>
            </div>
          </div>
          <div
            class="ant-space-item"
          >
            <div
              class="ant-table-wrapper"
            >
              <div
                class="ant-spin-nested-loading"
              >
                <div
                  class="ant-spin-container"
                >
                  <div
                    class="ant-table ant-table-small ant-table-bordered"
                  >
                    <div
                      class="ant-table-container"
                    >
                      <div
                        class="ant-table-content"
                      >
                        <table
                          style="table-layout: auto;"
                        >
                          <colgroup>
                            <col
                              style="width: 150px;"
                            />
                            <col
                              style="width: 150px;"
                            />
                            <col
                              style="width: 150px;"
                            />
                            <col />
                            <col
                              style="width: 150px;"
                            />
                          </colgroup>
                          <thead
                            class="ant-table-thead"
                          >
                            <tr>
                              <th
                                class="ant-table-cell ant-table-column-has-sorters"
                                tabindex="0"
                              >
                                <div
                                  class="ant-table-column-sorters"
                                >
                                  <span
                                    class="ant-table-column-title"
                                  >
                                    ID
                                  </span>
                                  <span
                                    class="ant-table-column-sorter ant-table-column-sorter-full"
                                  >
                                    <span
                                      class="ant-table-column-sorter-inner"
                                    >
                                      <span
                                        aria-label="caret-up"
                                        class="anticon anticon-caret-up ant-table-column-sorter-up"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="caret-up"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="0 0 1024 1024"
                                          width="1em"
                                        >
                                          <path
                                            d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                          />
                                        </svg>
                                      </span>
                                      <span
                                        aria-label="caret-down"
                                        class="anticon anticon-caret-down ant-table-column-sorter-down"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="caret-down"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="0 0 1024 1024"
                                          width="1em"
                                        >
                                          <path
                                            d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </span>
                                </div>
                              </th>
                              <th
                                class="ant-table-cell ant-table-column-has-sorters"
                                tabindex="0"
                              >
                                <div
                                  class="ant-table-column-sorters"
                                >
                                  <span
                                    class="ant-table-column-title"
                                  >
                                    Year
                                  </span>
                                  <span
                                    class="ant-table-column-sorter ant-table-column-sorter-full"
                                  >
                                    <span
                                      class="ant-table-column-sorter-inner"
                                    >
                                      <span
                                        aria-label="caret-up"
                                        class="anticon anticon-caret-up ant-table-column-sorter-up"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="caret-up"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="0 0 1024 1024"
                                          width="1em"
                                        >
                                          <path
                                            d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                          />
                                        </svg>
                                      </span>
                                      <span
                                        aria-label="caret-down"
                                        class="anticon anticon-caret-down ant-table-column-sorter-down"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="caret-down"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="0 0 1024 1024"
                                          width="1em"
                                        >
                                          <path
                                            d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </span>
                                </div>
                              </th>
                              <th
                                class="ant-table-cell ant-table-column-has-sorters"
                                tabindex="0"
                              >
                                <div
                                  class="ant-table-column-sorters"
                                >
                                  <span
                                    class="ant-table-column-title"
                                  >
                                    Semester
                                  </span>
                                  <span
                                    class="ant-table-column-sorter ant-table-column-sorter-full"
                                  >
                                    <span
                                      class="ant-table-column-sorter-inner"
                                    >
                                      <span
                                        aria-label="caret-up"
                                        class="anticon anticon-caret-up ant-table-column-sorter-up"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="caret-up"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="0 0 1024 1024"
                                          width="1em"
                                        >
                                          <path
                                            d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                          />
                                        </svg>
                                      </span>
                                      <span
                                        aria-label="caret-down"
                                        class="anticon anticon-caret-down ant-table-column-sorter-down"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="caret-down"
                                          fill="currentColor"
                                          focusable="false"
                                          height="1em"
                                          viewBox="0 0 1024 1024"
                                          width="1em"
                                        >
                                          <path
                                            d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </span>
                                </div>
                              </th>
                              <th
                                class="ant-table-cell ant-table-column-has-sorters"
                                tabindex="0"
                              >
                                <div
                                  class="ant-table-filter-column"
                                >
                                  <span
                                    class="ant-table-column-title"
                                  >
                                    <div
                                      class="ant-table-column-sorters"
                                    >
                                      <span
                                        class="ant-table-column-title"
                                      >
                                        Name
                                      </span>
                                      <span
                                        class="ant-table-column-sorter ant-table-column-sorter-full"
                                      >
                                        <span
                                          class="ant-table-column-sorter-inner"
                                        >
                                          <span
                                            aria-label="caret-up"
                                            class="anticon anticon-caret-up ant-table-column-sorter-up"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-up"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                          <span
                                            aria-label="caret-down"
                                            class="anticon anticon-caret-down ant-table-column-sorter-down"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-down"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                        </span>
                                      </span>
                                    </div>
                                  </span>
                                  <span
                                    class="ant-dropdown-trigger ant-table-filter-trigger"
                                    role="button"
                                    tabindex="-1"
                                  >
                                    <span
                                      aria-label="filter"
                                      class="anticon anticon-filter"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="filter"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="64 64 896 896"
                                        width="1em"
                                      >
                                        <path
                                          d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                        />
                                      </svg>
                                    </span>
                                  </span>
                                </div>
                              </th>
                              <th
                                class="ant-table-cell"
                              >
                                Action
                              </th>
                            </tr>
                          </thead>
                          <tbody
                            class="ant-table-tbody"
                          >
                            <tr
                              class="ant-table-row ant-table-row-level-0"
                              data-row-key="course_id"
                            >
                              <td
                                class="ant-table-cell"
                              >
                                course_id
                              </td>
                              <td
                                class="ant-table-cell"
                              >
                                2022
                              </td>
                              <td
                                class="ant-table-cell"
                              >
                                10
                              </td>
                              <td
                                class="ant-table-cell"
                              >
                                course1
                              </td>
                              <td
                                class="ant-table-cell"
                              >
                                <div
                                  class="ant-space ant-space-horizontal ant-space-align-center"
                                >
                                  <div
                                    class="ant-space-item"
                                  >
                                    <a
                                      href="/course/course_id/overview"
                                    >
                                      Go to
                                    </a>
                                  </div>
                                </div>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                  <ul
                    class="ant-pagination ant-pagination-mini ant-table-pagination ant-table-pagination-right"
                    unselectable="unselectable"
                  >
                    <li
                      aria-disabled="true"
                      class="ant-pagination-prev ant-pagination-disabled"
                      title="Previous Page"
                    >
                      <button
                        class="ant-pagination-item-link"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          aria-label="left"
                          class="anticon anticon-left"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="left"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                            />
                          </svg>
                        </span>
                      </button>
                    </li>
                    <li
                      class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                      tabindex="0"
                      title="1"
                    >
                      <a
                        rel="nofollow"
                      >
                        1
                      </a>
                    </li>
                    <li
                      aria-disabled="true"
                      class="ant-pagination-next ant-pagination-disabled"
                      title="Next Page"
                    >
                      <button
                        class="ant-pagination-item-link"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          aria-label="right"
                          class="anticon anticon-right"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="right"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                            />
                          </svg>
                        </span>
                      </button>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
